File: Sources/NumericSwiftWorkbenchKit/Domains/Integration.swift
//
//  Integration.swift  (Domains/)
//  NumericSwiftWorkbenchKit
//
//  Reference per-domain workbench suite: Integration (quadrature).
//
//  This is the pattern every other domain copies:
//    1. `<domain>Suite` — a `DomainSuite` naming the domain and wiring its
//       strategy + envelope registries.
//    2. `register<Domain>Strategies(into:)` — one closure per strategy id,
//       extracting the fixture `inputs` bag and delegating to NumericSwift.
//       Each closure forwards any ``NumericDiagnostic`` the library emitted.
//    3. `make<Domain>EnvelopeRegistry()` — per-(strategy, tier) accuracy bounds.
//
//  ## Self-awareness
//
//  Out-of-envelope cases are tagged `inEnvelope: false` in the fixture. For the
//  gate to pass, the *library* must emit an ``NumericDiagnostic/outsideEnvelope``
//  for those cases — the strategy closure only forwards what the library
//  produced, it never fabricates a diagnostic. `quad` emits one when its adaptive
//  subdivision limit is reached without meeting tolerance (SciPy-parity
//  `IntegrationWarning`).
//
//  Licensed under the Apache License, Version 2.0.
//

import Foundation
import NumericSwift

extension Workbench {

    /// The Integration (quadrature) domain suite.
    public static let integrationSuite = DomainSuite(
        name: "integration",
        registerStrategies: registerIntegrationStrategies,
        makeEnvelopeRegistry: makeIntegrationEnvelopeRegistry
    )
}

// MARK: - Integrand resolver

/// Resolve the integrand named by the fixture's `tag` input.
///
/// Closed-form integrands keep fixtures small: the JSON carries the tag and the
/// bounds, not a sampled function.
@Sendable
private func integrationIntegrand(_ tag: String?) -> (@Sendable (Double) -> Double)? {
    switch tag {
    case "gaussian_bell":   return { x in exp(-x * x) }
    case "sine":            return { x in sin(x) }
    case "cosine":          return { x in cos(x) }
    case "polynomial_deg2": return { x in x * x + 2 * x + 1 }     // (x+1)^2
    case "exp":             return { x in exp(x) }
    case "runge":           return { x in 1.0 / (1.0 + 25.0 * x * x) }
    case "oscillatory":     return { x in sin(50.0 * x) }
    // Strong endpoint singularity — used for out-of-envelope quad cases.
    case "inverse_sqrt":    return { x in x <= 0 ? 0 : 1.0 / sqrt(x) }
    default:                return nil
    }
}

// MARK: - Strategy registrations

/// Populate `registry` with the Integration (quadrature) strategies.
@Sendable
public func registerIntegrationStrategies(into registry: inout StrategyRegistry) {

    // quad — Gauss-Kronrod 15-point adaptive. Honours optional epsabs/epsrel/limit
    // inputs so out-of-envelope cases can force a tight subdivision budget.
    registry.register(id: "quad") { inputs in
        guard let a = inputs["a"]?.doubleValue,
              let b = inputs["b"]?.doubleValue,
              let f = integrationIntegrand(inputs["tag"]?.stringValue)
        else { return nil }
        let epsabs = inputs["epsabs"]?.doubleValue ?? quadDefaultEpsAbs
        let epsrel = inputs["epsrel"]?.doubleValue ?? quadDefaultEpsRel
        let limit = inputs["limit"]?.intValue ?? quadDefaultLimit
        let r = quad(f, a, b, epsabs: epsabs, epsrel: epsrel, limit: limit)
        return StrategyResult(value: r.value, diagnostics: r.diagnostics)
    }

    // romberg — Richardson-extrapolation quadrature.
    registry.register(id: "romberg") { inputs in
        guard let a = inputs["a"]?.doubleValue,
              let b = inputs["b"]?.doubleValue,
              let f = integrationIntegrand(inputs["tag"]?.stringValue)
        else { return nil }
        let r = romberg(f, a, b)
        return StrategyResult(value: r.value, diagnostics: r.diagnostics)
    }

    // simps — composite Simpson's rule over n+1 samples.
    registry.register(id: "simps") { inputs in
        guard let (xs, ys) = sampledGrid(inputs) else { return nil }
        return StrategyResult(value: simps(ys, x: xs))
    }

    // trapz — composite trapezoidal rule over n+1 samples.
    registry.register(id: "trapz") { inputs in
        guard let (xs, ys) = sampledGrid(inputs) else { return nil }
        return StrategyResult(value: trapz(ys, x: xs))
    }

    // fixed_quad — Gauss-Legendre fixed-order quadrature.
    registry.register(id: "fixed_quad") { inputs in
        guard let a = inputs["a"]?.doubleValue,
              let b = inputs["b"]?.doubleValue,
              let f = integrationIntegrand(inputs["tag"]?.stringValue)
        else { return nil }
        let order = inputs["order"]?.intValue ?? 5
        return StrategyResult(value: fixedQuad(f, a, b, n: order))
    }
}

/// Build an `(xs, ys)` sample grid for the sample-based strategies (simps/trapz).
///
/// Reads `a`, `b`, `n` (number of subintervals) and the integrand `tag`.
@Sendable
private func sampledGrid(_ inputs: [String: InputValue]) -> ([Double], [Double])? {
    guard let a = inputs["a"]?.doubleValue,
          let b = inputs["b"]?.doubleValue,
          let n = inputs["n"]?.intValue, n > 0,
          let f = integrationIntegrand(inputs["tag"]?.stringValue)
    else { return nil }
    let h = (b - a) / Double(n)
    let xs = (0...n).map { a + Double($0) * h }
    return (xs, xs.map(f))
}

// MARK: - Envelope registry

/// Per-(strategy, tier) accuracy envelopes for the Integration domain.
///
///  - `quad` (Gauss-Kronrod adaptive): full double precision on smooth integrands.
///  - `romberg` (Richardson extrapolation): looser on non-smooth integrands.
///  - `simps` (Simpson's rule): exact for polynomials ≤ degree 3.
///  - `trapz` (trapezoidal rule): first-order; exact only for linear functions.
///  - `fixed_quad` (Gauss-Legendre fixed-point): order-dependent.
@Sendable
public func makeIntegrationEnvelopeRegistry() -> EnvelopeRegistry {
    var reg = EnvelopeRegistry()
    for tier: CaseTier in [.trivial, .hard, .edge] {
        let quadTol: Double = tier == .trivial ? 1e-12 : tier == .hard ? 1e-8 : 1e-5
        reg.register(EnvelopeEntry(strategy: "quad", tier: tier, maxAbsError: quadTol,
            description: "Gauss-Kronrod 15-point adaptive quadrature — \(tier.rawValue) cases"))

        let rombergTol: Double = tier == .trivial ? 1e-10 : tier == .hard ? 1e-6 : 1e-3
        reg.register(EnvelopeEntry(strategy: "romberg", tier: tier, maxAbsError: rombergTol,
            description: "Romberg (Richardson extrapolation) — \(tier.rawValue) cases"))

        let simpsTol: Double = tier == .trivial ? 1e-9 : tier == .hard ? 1e-3 : 1e-1
        reg.register(EnvelopeEntry(strategy: "simps", tier: tier, maxAbsError: simpsTol,
            description: "Composite Simpson's rule — \(tier.rawValue) cases"))

        let trapzTol: Double = tier == .trivial ? 1e-6 : tier == .hard ? 1e-1 : 1e0
        reg.register(EnvelopeEntry(strategy: "trapz", tier: tier, maxAbsError: trapzTol,
            description: "Composite trapezoidal rule — \(tier.rawValue) cases"))

        let fixedTol: Double = tier == .trivial ? 1e-9 : tier == .hard ? 1e-4 : 1e-2
        reg.register(EnvelopeEntry(strategy: "fixed_quad", tier: tier, maxAbsError: fixedTol,
            description: "Gauss-Legendre fixed quadrature — \(tier.rawValue) cases"))
    }
    return reg
}

File: Tools/workbench_oracles/integration.py
#!/usr/bin/env python3
"""Oracle generator for the NumericSwift workbench — Integration (quadrature) domain.

Computes bit-exact reference integrals with scipy and freezes them as the JSON
fixture `Tests/NumericSwiftTests/Fixtures/workbench/integration.json`.

This is the REFERENCE generator every other domain copies. Contract (WORKBENCH.md
§2/§3/§5):

  * ~100 cases partitioned 10 / 80 / 10 (trivial / hard / edge); the actual counts
    are printed so a thin tier is never silently shipped.
  * Each case carries `oracle.bits` (IEEE-754 hex) as the canonical value and a
    `source` citation. Oracle values come ONLY from scipy — never from NumericSwift
    (FP3 vacuous-gate rule).
  * `inEnvelope` is per-strategy. Out-of-envelope cases (here: `quad` forced to a
    tiny subdivision limit on a hard integrand) are tagged `false`, so the gate
    requires NumericSwift to emit an `outsideEnvelope` diagnostic for them.

Integrand `tag`s MUST match `integrationIntegrand` in
`Sources/NumericSwiftWorkbenchKit/Domains/Integration.swift`.

Run:
    /tmp/.nsoracle/bin/python Tools/workbench_oracles/integration.py
"""

import json
import math
import struct
import sys
from pathlib import Path

import numpy as np
from scipy import integrate

SOURCE = f"scipy.integrate.quad {__import__('scipy').__version__}"

# Integrands — keep in lockstep with the Swift `integrationIntegrand` resolver.
INTEGRANDS = {
    "gaussian_bell": lambda x: math.exp(-x * x),
    "sine": math.sin,
    "cosine": math.cos,
    "polynomial_deg2": lambda x: x * x + 2 * x + 1,
    "exp": math.exp,
    "runge": lambda x: 1.0 / (1.0 + 25.0 * x * x),
    "oscillatory": lambda x: math.sin(50.0 * x),
    "inverse_sqrt": lambda x: 0.0 if x <= 0 else 1.0 / math.sqrt(x),
}


def bits_hex(value: float) -> str:
    """IEEE-754 bit pattern of a double as a 0x-prefixed 16-hex-digit string."""
    return "0x%016X" % struct.unpack("<Q", struct.pack("<d", value))[0]


def oracle(tag: str, a: float, b: float) -> float:
    """High-accuracy reference integral from scipy."""
    f = INTEGRANDS[tag]
    val, _ = integrate.quad(f, a, b, limit=200, epsabs=1e-13, epsrel=1e-13)
    return val


def case(
    cid,
    tier,
    tag,
    a,
    b,
    strategies,
    tol,
    *,
    n=None,
    order=None,
    limit=None,
    epsabs=None,
    epsrel=None,
    in_envelope=None,
):
    inputs = {"a": a, "b": b, "tag": tag}
    if n is not None:
        inputs["n"] = n
    if order is not None:
        inputs["order"] = order
    if limit is not None:
        inputs["limit"] = limit
    if epsabs is not None:
        inputs["epsabs"] = epsabs
    if epsrel is not None:
        inputs["epsrel"] = epsrel
    val = oracle(tag, a, b)
    entry = {
        "id": cid,
        "tier": tier,
        "inputs": inputs,
        "oracle": {"value": val, "bits": bits_hex(val)},
        "source": SOURCE,
        "strategies": strategies,
        "tol": tol,
    }
    if in_envelope:
        entry["inEnvelope"] = in_envelope
    return entry


def build():
    cases = []
    ALL = ["quad", "romberg", "simps", "trapz", "fixed_quad"]
    # Realistic per-strategy envelopes for smooth hard cases. quad/romberg are
    # near machine precision; simps is 4th-order; trapz is 1st-order; fixed_quad
    # is a single fixed-order Gauss-Legendre rule (no adaptivity) so it gets a
    # wide envelope on broad ranges. These are the declared envelopes, not the
    # achieved errors — the report's `error` column still shows what each hits.
    smooth_tol = {
        "quad": 1e-7,
        "romberg": 1e-4,
        "simps": 1e-2,
        "trapz": 5e0,
        "fixed_quad": 1e0,
    }

    # ── Trivial (~10): closed-form, smooth, fine grids ────────────────────────
    trivial = [
        ("polynomial_deg2", 0.0, 1.0),
        ("polynomial_deg2", -1.0, 2.0),
        ("sine", 0.0, math.pi),
        ("cosine", 0.0, math.pi / 2),
        ("exp", 0.0, 1.0),
        ("exp", -1.0, 1.0),
        ("sine", 0.0, 2 * math.pi),
        ("polynomial_deg2", 0.0, 5.0),
        ("cosine", -math.pi / 2, math.pi / 2),
        ("exp", 0.0, 2.0),
    ]
    for i, (tag, a, b) in enumerate(trivial):
        # trivial tier: tight tolerances, fine grid (n=1000)
        cases.append(
            case(
                f"integration.trivial.{tag}_{i}",
                "trivial",
                tag,
                a,
                b,
                ALL,
                {
                    "quad": 1e-12,
                    "romberg": 1e-10,
                    "simps": 1e-9,
                    "trapz": 1e-4,
                    "fixed_quad": 1e-6,
                },
                n=1000,
                order=8,
            )
        )

    # ── Hard (~80): realistic, varied tags / ranges / grids ───────────────────
    hard_tags = ["gaussian_bell", "runge", "sine", "cosine", "exp", "oscillatory"]
    ranges = [(-3.0, 3.0), (-1.0, 1.0), (0.0, 2.0), (-2.0, 5.0), (0.5, 4.0)]
    n_choices = [200, 400, 800]
    idx = 0
    for tag in hard_tags:
        for a, b in ranges:
            for n in n_choices:
                if idx >= 80:
                    break
                # Oscillatory / Runge integrands defeat the low-order and
                # fixed-order rules: widen their envelopes accordingly.
                tol = dict(smooth_tol)
                if tag in ("oscillatory", "runge"):
                    tol["romberg"] = 1e0
                    tol["simps"] = 1e0
                    tol["trapz"] = 1e1
                    tol["fixed_quad"] = 1e1
                cases.append(
                    case(
                        f"integration.hard.{tag}_{idx}",
                        "hard",
                        tag,
                        a,
                        b,
                        ALL,
                        tol,
                        n=n,
                        order=8,
                    )
                )
                idx += 1

    # ── Edge (~10): out-of-envelope quad (tiny limit) + hard in-envelope ──────
    # Out-of-envelope: quad with a 3-subdivision budget on hard integrands MUST
    # trip the IntegrationWarning-equivalent diagnostic. Run quad only.
    # Asymmetric / singular only: a symmetric oscillatory range lets gk15's
    # symmetric nodes cancel, yielding a tiny error estimate that is accepted
    # without subdivision (no diagnostic). Endpoint singularities and asymmetric
    # high-frequency integrands reliably exhaust the 3-subdivision budget.
    edge_oov = [
        ("oscillatory", 0.0, 10.0),
        ("inverse_sqrt", 0.0, 1.0),
        ("inverse_sqrt", 0.0, 4.0),
        ("inverse_sqrt", 0.0, 9.0),
        ("oscillatory", 0.3, 12.0),
    ]
    for i, (tag, a, b) in enumerate(edge_oov):
        cases.append(
            case(
                f"integration.edge.quad_oov_{i}",
                "edge",
                tag,
                a,
                b,
                ["quad"],
                {"quad": 1e-5},
                limit=3,
                in_envelope={"quad": False},
            )
        )

    # In-envelope edge: quad with a generous budget stays reliable, no diagnostic.
    edge_ok = [
        ("gaussian_bell", -8.0, 8.0),
        ("runge", -10.0, 10.0),
        ("exp", -3.0, 3.0),
        ("sine", 0.0, 8 * math.pi),
        ("cosine", 0.0, 8 * math.pi),
    ]
    for i, (tag, a, b) in enumerate(edge_ok):
        # quad only: this block is the false-positive guard — quad with a generous
        # budget stays inside its envelope and must NOT emit a diagnostic.
        cases.append(
            case(
                f"integration.edge.quad_ok_{i}",
                "edge",
                tag,
                a,
                b,
                ["quad"],
                {"quad": 1e-6},
            )
        )

    return cases


def main():
    cases = build()
    tiers = {"trivial": 0, "hard": 0, "edge": 0}
    for c in cases:
        tiers[c["tier"]] += 1
    print(
        f"integration: {len(cases)} cases — "
        f"trivial={tiers['trivial']} hard={tiers['hard']} edge={tiers['edge']}",
        file=sys.stderr,
    )

    out_dir = (
        Path(__file__).resolve().parents[2]
        / "Tests/NumericSwiftTests/Fixtures/workbench"
    )
    out_dir.mkdir(parents=True, exist_ok=True)
    out = out_dir / "integration.json"
    out.write_text(json.dumps(cases, indent=2) + "\n")
    print(f"wrote {out}", file=sys.stderr)


if __name__ == "__main__":
    main()
